Use builder pattern to simplify requests to external endpoints

JoelSpeed opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->
Creates a new request builder that simplifies fetching and parsing JSON requests to reduce duplication of the request handling across the project.

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Should simplify the code a little and make it more maintainable by adding an extensible builder pattern.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Unit testing all working. Basic manual test of OIDC provider.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

NickMeves commented on pkg/requests/builder.go:
```diff
@@ -0,0 +1,173 @@
+// Into performs the request and attempts to unmarshal the response into the
+// the given interface. The response body is assumed to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func (r *builder) Into(into interface{}) error {
```
Should the Into/JSON method variants be aligned to the `Unmarshal` naming scheme that encoding libraries use (and that you are doing under the hood)?

NickMeves replied:
I wonder if there's a clean way to register the target interface to Unmarshal into (or multiples...?) as a builder modifier in the chain (so the chain finishes with the `Do/Send` verb).  Error handling might get hairy trying to do it that way though.  (I'm just thinking out loud).

JoelSpeed replied:
We could change this up to have `Do` return a `Result`, that result could then be reused a few times if you wanted to `Into` a couple of things, but I haven't really seen any use case for that in our code at present.

`UnmarshalJSON` and `UnmarshalInto` make sense though

NickMeves replied:
I came across this dated project that tried to do HTTP request via builder pattern.  They seem to have implemented better than I could explain about having the unmarshal destination being a builder modifier method before the final execution:

https://github.com/BrandonRomano/wrecker

https://github.com/BrandonRomano/wrecker/blob/d37dcbf1a77802e496cddbcc55e28e5ee9583459/wrecker.go#L118

Could eliminate the need for the standalone sermi-duplicated `Into` that operated on a response body in the cases where you needed to evaluate the response & unmarshal into a struct -- would result in always getting both back if an `UnmarshalInto` or something was tagged on the builder.

Not sure how this fits with the `simplejson` output case though.

JoelSpeed replied:
I don't know how I feel about this. I don't think we could implement the simplejson method doing it that way so then we would have two patterns. What you you think of Do returning a `Result` which is a struct containing the response and a possible error, that could then have methods `UnmarshalInto` and `UnmarshalJSON` on them? I think I'd prefer that over the wrecker approach personally.

I think there were only two cases where we go the response and then unmarshalled into an object, I'm trying to think what this approach would save us, 1 error check?

NickMeves replied:
I like that idea a lot, seems like a good way to simplify the chaining while supporting simplejson as an output option.

The multiple executors per output format felt weird to me.  This route have a consistent executor `Do` that can be chained into a transformer `UnmarshalInto` and `UnmarshalJSON` feels very clean.

NickMeves replied:
What about access to the original error when `Do()` is chained off of by the transformer? Should `UnmarshalInto` and `UnmarshalJSON` return the original `Result` + unmarshal error so that the `Result.Error` field is accessible after the transformation attempt?  Or `UnmarshalInto` / `UnmarshalJSON` error immediately when `Result.Error` is non-Nil and only return `error` like normal unmarshal methods?

Chaining is so much harder in Golang with error outputs.  Much easier in other languages when the middle parts just throw a bazillion potential exceptions that developers frequently forget to handle correctly 😄

JoelSpeed replied:
Great, I'll have a go at that.

And I'd expect the unmarshal methods to look pretty much identical to now, except they'd check for an error in the result and return that early if it's there

NickMeves commented on pkg/requests/builder.go:
```diff
@@ -0,0 +1,173 @@
+// JSON performs the request and attempts to unmarshal the response into a
+// simplejson.Json. The response body is assume to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func (r *builder) JSON() (*simplejson.Json, error) {
```
Related to to the `unmarshal` comment, should this be renamed to be clearer and in "verb" form to convey it is intended as a terminal method in the builder chain?

Otherwise cursory glance at the API might lead someone to thing `JSON()` sets up an encoding format or something and needs an action method chained after it.

JoelSpeed replied:
Yeah, this one should definitely be `UnmarshalJSON` at a minimum

NickMeves commented on pkg/requests/builder.go:
```diff
@@ -0,0 +1,173 @@
+// Do performs the request and returns the response in its raw form.
+// If the request has already been performed, returns the previous result.
+// This will not allow you to repeat a request.
+func (r *builder) Do() (*http.Response, error) {
```
I'm torn on this: now after looking I see the naming aligns to `http.DefaultClient.Do(req)` -- but on a cursory glance at the PR the purpose of this function didn't jump out to me by eyeballing the name.

"Do... what?" was my first thought.  Is `Send` potentially more clear?

JoelSpeed replied:
I think this becomes move obvious when it's used, `requests.New(url.String()).Do()` is a pattern I've seen before, it was somewhere loosely based on the Kubernetes REST client  https://github.com/kubernetes/client-go/blob/1e6150831a960cf38075bb8561e01f48837f8d74/rest/request.go#L952

`Send` to me feels a bit wrong 😅 It sends but also receives the result of the request 🤔

NickMeves replied:
It looks like `Do` seems to be an overwhelming convention in this space -- lets stick with that! Today I Learned 😄

NickMeves commented on pkg/requests/builder.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+// WithHeader replaces the request header map with the given header map.
+func (r *builder) WithHeader(header http.Header) Builder {
```
Another style preference question: I see the alignment to the singular naming of multiple headers in `http.Header` -- but the distinction of this setting all headers via a map vs the `SetHeader` function immediately below this setting a single header made it harder to tell `WithHeader` is intended to set all headers in bulk.

`WithHeaders` might be clearer for the API contract you are building, but then it deviates from the underlying source conventions.

JoelSpeed replied:
What about `WithHeaderMap`? Would that make sense?

NickMeves commented on pkg/requests/builder.go:
```diff
@@ -0,0 +1,173 @@
+// if the response was successful.
+func getResponseBody(resp *http.Response) ([]byte, error) {
+	body, err := ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
```
Do we want to handle this error? I've had it trained into me to handle all errors (my IDE likes to yell at me about it) -- but I even seen go underlying `http` library ignoring error handling on Body.Close()

JoelSpeed replied:
Had a dig through the code, given we have done an `ioutil.ReadAll` just above, there's no need to check the error as the code path checks whether the body has been read, because it is at EOF, it returns a non-nil error. What we should do is check the error from the line above first, the typical pattern is to defer the response body close so I'll do that

JoelSpeed commented:
I've renamed `Into` and `JSON` to prefix them both with `Unmarshal`, moved the `resp.Body.Close()` to a defer as is idomatic, and added a changelog

Outstanding comments are whether to rename `WithHeader`, `WithHeaders` or `WithHeaderMap` is fine by me, and whether to rename `Do`, I prefer `Do` to `Send`, third contended, `Complete`? (Why is naming so hard...)

NickMeves commented:
> I've renamed `Into` and `JSON` to prefix them both with `Unmarshal`, moved the `resp.Body.Close()` to a defer as is idomatic, and added a changelog
> 
> Outstanding comments are whether to rename `WithHeader`, `WithHeaders` or `WithHeaderMap` is fine by me, and whether to rename `Do`, I prefer `Do` to `Send`, third contended, `Complete`? (Why is naming so hard...)

Hardest things in CS, caching & naming things 😄 

I think either `WithHeaders` or `WithHeaderMap` are good -- more clear that it isn't operating on a single header.

JoelSpeed commented:
@NickMeves I've updated this so it now uses a Do -> Result pattern, also renamed `WithHeader` to `WithHeaders`, I know that doesn't match the `http.Header` struct but I think it's ok, please take another look

JoelSpeed commented:
Had an accidental focused test so the tests failed, fixed now, @NickMeves are you around to re-tick?
